src/main/java/com/squareup/javapoet/SelfName.java

```diff
@@ -0,0 +1,26 @@
+package com.squareup.javapoet;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+
+public class SelfName extends TypeName {
+
+  public static SelfName get() {
+    return new SelfName(Collections.<AnnotationSpec>emptyList());
+  }
+
+  SelfName(List<AnnotationSpec> annotations) {
+    super(annotations);
+  }
+
+  @Override public TypeName annotated(List<AnnotationSpec> annotations) {
+    return new SelfName(annotations);
+  }
+
+  @Override CodeWriter emit(CodeWriter out) throws IOException {
+    ClassName name = out.resolve(out.peekType().name);
+    return emitAnnotations(out).emitAndIndent(out.lookupName(name));
```

Comment:
This feels like a circular call might arise. `lookupName` uses `resolve` internally. Can you think of a way simplifying this? I mean, get a `ClassName` instance from peeked `TypeName`.